File: packages/flutter_fuels/lib/fuel_wallet.dart
import 'platform_impl/stub_wallet.dart'
    if (dart.library.io) 'platform_impl/mobile_wallet.dart'
    if (dart.library.html) 'platform_impl/web_wallet.dart';

class FuelWallet {
  final String? mnemonicPhrase;
  final String bech32Address;
  final String b256Address;
  final String privateKey;
  final dynamic networkProvider;

  const FuelWallet({
    required this.bech32Address,
    required this.b256Address,
    required this.mnemonicPhrase,
    required this.privateKey,
    required this.networkProvider,
  });

  static final _wallet = FuelWalletImpl();

  static Future<FuelWallet> generateNewWallet({
    required dynamic networkProvider,
  }) async {
    final data = await _wallet.generateNewWallet();

    return FuelWallet(
      bech32Address: data['address']['bech32Address'],
      b256Address: data['address']['b256Address'],
      mnemonicPhrase: data['mnemonicPhrase'],
      privateKey: data['privateKey'],
      networkProvider: networkProvider,
    );
  }

  static Future<FuelWallet> newFromPrivateKey({
    required dynamic networkProvider,
    required String privateKey,
  }) async {
    final data = await _wallet.newFromPrivateKey(
      privateKey: privateKey,
    );

    return FuelWallet(
      bech32Address: data['address']['bech32Address'],
      b256Address: data['address']['b256Address'],
      mnemonicPhrase: data['mnemonicPhrase'],
      privateKey: data['privateKey'],
      networkProvider: networkProvider,
    );
  }

  static Future<FuelWallet> newFromMnemonicPhrase({
    required dynamic networkProvider,
    required String mnemonic,
  }) async {
    final data = await _wallet.newFromMnemonic(
      mnemonic: mnemonic,
    );

    return FuelWallet(
      bech32Address: data['address']['bech32Address'],
      b256Address: data['address']['b256Address'],
      mnemonicPhrase: data['mnemonicPhrase'],
      privateKey: data['privateKey'],
      networkProvider: networkProvider,
    );
  }

  Future<String> transfer({
    required String destinationB256Address,
    required int fractionalAmount,
    required String assetId,
    required int gasPrice,
    required int gasLimit,
    required int maturity,
  }) async {
    return _wallet.transfer(
      networkProvider: networkProvider,
      privateKey: privateKey,
      destinationB256Address: destinationB256Address,
      fractionalAmount: fractionalAmount,
      assetId: assetId,
      gasPrice: gasPrice,
      gasLimit: gasLimit,
      maturity: maturity,
    );
  }

  Future<String> signMessage({
    required String message,
  }) async {
    return _wallet.signMessage(
      networkProvider: networkProvider,
      privateKey: privateKey,
      message: message,
    );
  }

  Future<String> sendTransaction({
    required dynamic transactionRequest,
  }) async {
    return _wallet.sendTransaction(
      networkProvider: networkProvider,
      privateKey: privateKey,
      transactionRequest: transactionRequest,
    );
  }
}

File: packages/flutter_fuels/web/src/fuel_wallet.ts
import {
  Mnemonic,
  Provider,
  ScriptTransactionRequest,
  serializeTransactionResponseJson,
  TransactionBlob,
  TransactionCreate,
  TransactionRequest,
  transactionRequestify,
  TransactionRequestLike,
  TransactionScript,
  TransactionUpgrade,
  TransactionUpload,
  Wallet,
  WalletUnlocked,
} from "fuels";

type AddressObject = {
  b256Address: string;
};

type WalletObject = {
  address: AddressObject;
  privateKey: string;
  mnemonicPhrase?: string;
};

function walletToJson(wallet: WalletUnlocked, mnemonic?: string): WalletObject {
  return {
    address: {
      b256Address: wallet.address.toB256(),
    },
    privateKey: wallet.privateKey,
    mnemonicPhrase: mnemonic,
  };
}

class WalletInterface {
  async generateNewWallet(): Promise<WalletObject> {
    let mnemonic = Mnemonic.generate(16);
    return this.newWalletFromMnemonic(mnemonic);
  }

  async newWalletFromMnemonic(mnemonic: string): Promise<WalletObject> {
    let wallet = Wallet.fromMnemonic(mnemonic);
    return walletToJson(wallet, mnemonic);
  }

  async newWalletFromMnemonicAndPath(mnemonic: string, path: string): Promise<WalletObject> {
    let wallet = Wallet.fromMnemonic(mnemonic, path);
    return walletToJson(wallet, mnemonic);
  }

  async newWalletFromPrivateKey(privateKey: string): Promise<WalletObject> {
    let wallet = Wallet.fromPrivateKey(privateKey);
    return walletToJson(wallet);
  }

  async signMessage(
    privateKey: string,
    message: string
  ): Promise<string> {
    let wallet = Wallet.fromPrivateKey(privateKey);
    return await wallet.signMessage(message);
  }

  async sendTransaction(
    privateKey: string,
    networkUrl: string,
    transactionRequestJson: string,
    returnTransactionResponse: boolean
  ): Promise<string> {
    let provider = new Provider(networkUrl);
    let wallet = Wallet.fromPrivateKey(privateKey, provider);
    let transactionRequest = JSON.parse(transactionRequestJson);
    let response = await wallet.sendTransaction(transactionRequest);
    let txResult = await response.waitForResult()
    if (!txResult.isStatusSuccess) {
      throw new Error(`Transaction failed: ${txResult.status}`);
    }
    if (!returnTransactionResponse) {
      return response.id;
    }
    try {
      const serializedTxResponse = await serializeTransactionResponseJson(response);
      return JSON.stringify(serializedTxResponse);
    } catch (error) {
      console.error('Error serializing transaction response:', error);
    }
    return response.id;
  }

  async simulateTransaction(
    privateKey: string,
    networkUrl: string,
    transactionRequestJson: string
  ): Promise<string> {
    let provider = new Provider(networkUrl);
    let wallet = Wallet.fromPrivateKey(privateKey, provider);
    let transactionRequest = JSON.parse(transactionRequestJson);
    let response = await wallet.simulateTransaction(transactionRequest);

    return JSON.stringify(response);
  }

  async genTransferTransactionRequest(
    privateKey: string,
    networkUrl: string,
    destinationB256Address: string,
    fractionalAmount: number,
    assetId: string,
  ): Promise<[string, string]> {
    let provider = new Provider(networkUrl);
    let wallet = Wallet.fromPrivateKey(privateKey, provider);

    let request = new ScriptTransactionRequest({gasLimit: 20000});
    let isUserAccount = await provider.isUserAccount(destinationB256Address);
    if (!isUserAccount) {
      throw new Error(`Not a valid user address`);
    }
    request = wallet.addTransfer(request, {destination: destinationB256Address, amount: fractionalAmount, assetId});
    const txCost = await wallet.getTransactionCost(request);
    request = await wallet.fund(request, txCost);

    const txId = request.getTransactionId(await provider.getChainId())
    return [JSON.stringify(request), txId];
  }
}

class FuelsUtils {
  transformTxRequest(transactionRequestJson: string): string {
    let txRequestLike: TransactionRequestLike = JSON.parse(transactionRequestJson);
    let txRequest: TransactionRequest = transactionRequestify(txRequestLike);
    let tx: TransactionBlob | TransactionScript | TransactionCreate | TransactionUpgrade | TransactionUpload = txRequest.toTransaction();
    return JSON.stringify(tx);
  }

  async getTransactionCost(
    networkUrl: string,
    transactionRequestJson: string
  ): Promise<string> {
    let provider = new Provider(networkUrl);
    let transactionRequest = JSON.parse(transactionRequestJson);
    const {gasPrice, gasUsed, minFee, maxFee, minGas, maxGas} = await provider.getTransactionCost(transactionRequest)
    const responseObject = {
      gasPrice: gasPrice.toNumber(),
      gasUsed: gasUsed.toNumber(),
      minFee: minFee.toNumber(),
      maxFee: maxFee.toNumber(),
      minGas: minGas.toNumber(),
      maxGas: maxGas.toNumber(),
    }
    return JSON.stringify(responseObject);
  }

  async isUserAccount(
    networkUrl: string,
    address: string
  ): Promise<boolean> {
    let provider = new Provider(networkUrl);
    return provider.isUserAccount(address);
  }
}

function injectObject(target: object, obj: object, name: PropertyKey) {
  Object.defineProperty(target, name, {
    value: obj,
    writable: false,
    enumerable: true,
    configurable: false,
  });

  if (typeof document !== "undefined") {
    const loadedEvent = new CustomEvent(`${String(name)} loaded`, {
      detail: obj,
    });
    document.dispatchEvent(loadedEvent);
  }
}

injectObject(window, new WalletInterface(), "flutter_fuels_wallet");
injectObject(window, new FuelsUtils(), "flutter_fuels_utils");
